### PR TITLE
Update `list_jobs` function in `DatabricksHook` to token-based pagination 

### PR DESCRIPTION
**Pull request by oleksiidav**
Update Airflow to switch from default offset-based pagination (deprecated) to using a page token

### CODE REVIEW

File: airflow/providers/databricks/hooks/databricks.py [hunk 1]
```diff
@@ -156,13 +156,13 @@ def submit_run(self, json: dict) -> int:
         return response["run_id"]
 
     def list_jobs(
-        self, limit: int = 25, offset: int = 0, expand_tasks: bool = False, job_name: str | None = None
+        self, limit: int = 25, page_token: str | None = None, expand_tasks: bool = False, job_name: str | None = None
```
Comment (eladkal) [inline airflow/providers/databricks/hooks/databricks.py]: this is a breaking change. We should deprecate first.

**Review by eladkal — CHANGES_REQUESTED** [review]

**Comment by eladkal** [comment]
What is `JOBS-13077` ?

File: airflow/providers/databricks/hooks/databricks.py (hunk 1 above)
Comment (alexott) [reply airflow/providers/databricks/hooks/databricks.py]: Are people supposed to use hook directly, especially for such advanced cases like offsets?

**Comment by oleksiidav** [comment]
> What is `JOBS-13077` ?

Tracking ticket from Databricks side, nevermind

File: airflow/providers/databricks/hooks/databricks.py (hunk 1 above)
Comment (oleksiidav) [reply airflow/providers/databricks/hooks/databricks.py]: It has been deprecated in [Databricks API docs](https://docs.databricks.com/api/workspace/jobs/list) for a few months now if that helps

Comment (oleksiidav) [reply airflow/providers/databricks/hooks/databricks.py]: Also, Databricks has already gone through the breaking change process to communicate to the customers who are using the JobsList API with offsets (through Airflow), urging them to update ASAP @eladkal

Comment (eladkal) [reply airflow/providers/databricks/hooks/databricks.py]: If both `offset` and `page_token` are supported by the API then we need to support them both with raising deprecation warning when using `offset`

Comment (oleksiidav) [reply airflow/providers/databricks/hooks/databricks.py]: @eladkal Done, enabled both with a deprecation warning, PTAL

**Review by alexott — CHANGES_REQUESTED** [review]
code looks good - can we add a unit test for token usage?

**Comment by oleksiidav** [comment]
Added test @alexott, PTAL!

**Review by alexott — APPROVED** [review]
lgtm.

@eladkal WDYT?

**Comment by oleksiidav** [comment]
Hi @eladkal, could you please take another look?

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
LGTM @eladkal ?

File: airflow/providers/databricks/hooks/databricks.py [hunk 2]
```diff
@@ -156,26 +156,37 @@ def submit_run(self, json: dict) -> int:
         return response["run_id"]
 
     def list_jobs(
-        self, limit: int = 25, offset: int = 0, expand_tasks: bool = False, job_name: str | None = None
+        self, limit: int = 25, offset: int | None = None, page_token: str | None = None, expand_tasks: bool = False, job_name: str | None = None
```
Comment (eladkal) [inline airflow/providers/databricks/hooks/databricks.py]: please add the new parameter last (to avoid breaking code of users who use positional arguments)

Comment (oleksiidav) [reply airflow/providers/databricks/hooks/databricks.py]: Done!

**Review by eladkal — APPROVED** [review]

File: airflow/providers/databricks/hooks/databricks.py [hunk 3]
```diff
@@ -165,17 +165,28 @@ def list_jobs(
...
         :param job_name: Optional name of a job to search.
+        :param page_token: The optional page token pointing at the first first job to return.
         :return: A list of jobs.
         """
         has_more = True
         all_jobs = []
-
+        use_token_pagination = (page_token is not None) or (offset is None)
+        if offset is not None:
+            print("[WARN] You are using the offset parameter in list_jobs which will be deprecated soon")
```
Comment (eladkal) [inline airflow/providers/databricks/hooks/databricks.py]: What does it mean soon?
How do we know this?

If this is Airflow deprecation we should use `AirflowProviderDeprecationWarning`. if this is databricks deprecation we should not say anything about it.

**Review by eladkal — CHANGES_REQUESTED** [review]

File: airflow/providers/databricks/hooks/databricks.py (hunk 3 above)
Comment (oleksiidav) [reply airflow/providers/databricks/hooks/databricks.py]: It follows from [this discussion](https://github.com/apache/airflow/pull/33472#discussion_r1297282608) - the parameter has been deprecated by Databricks already, and the plan would be to show a deprecation warning in Airflow once this PR is merged, and then remove it from the hook altogether after 2 releases (around 1 month).

If we don't do this by Oct, and the Airflow hook still uses the offset param, then Databricks-enforced API limits will kick in, resulting in many offset-based requests possibly failing, which would be a worse way of handling a breaking change than deprecating it softly with an intermediate warning. Let's drive the switch to token-based pagination asap.

Comment (potiuk) [reply airflow/providers/databricks/hooks/databricks.py]: Let's write what you wrote more expiicitly in the message, including the cut-off date.

Comment (oleksiidav) [reply airflow/providers/databricks/hooks/databricks.py]: Done! @eladkal PTAL

**Comment by oleksiidav** [comment]
Heyy @eladkal , I have updated the tests/formatting, can we pls merge?

**Review by eladkal — APPROVED** [review]

**Comment by eladkal** [comment]
@oleksiidav static checks are failing you are missing import of `AirflowProviderDeprecationWarning`

**Comment by oleksiidav** [comment]
Thanks @eladkal @potiuk ! Fixed it now. Did not have access to test/linter workflow to run checks

**Pull request merged**